Blocking KokoroEngine

martindevans opened the pull request:
Modified KokoroEngine to block, instead of poll with a very short timer.

This should be more responsive (no longer waiting 10ms between every job) and less wasteful (no longer spinning when waiting for work).

Since there's a cancellation token a future PR could wire it up into `kokoroJob.Progress(model);` to cancel that work partway, instead of it being an atomic unit.

Lyrcaxis commented on Core/KokoroEngine.cs:
```diff
@@ -8,42 +8,36 @@
-    volatile bool hasExited;
+    protected readonly BlockingCollection<KokoroJob> queuedJobs = [];
+    private readonly CancellationTokenSource _cancellation = new();
 
     /// <summary> Creates a new Kokoro Engine instance, loading the model into memory and initializing a background worker thread to continuously scan for newly queued jobs, dispatching them in order, when it's free. </summary>
     /// <remarks> If 'options' is specified, the model will be loaded with them. This is particularly useful when needing to run on non-CPU backends, as the default backend otherwise is the CPU with 8 threads. </remarks>
-    public KokoroEngine(string modelPath, SessionOptions options = null) {
+    public KokoroEngine(string modelPath, SessionOptions options = null)
+    {
```
This repo's code style uses same-line brackets and ommited private modifier without `_`.

Lyrcaxis reviewed (changes requested):
I like the idea of blocking vs polling, and if it would simplify the code it'd be further nice.

Needs some tweaks to keep existing functionality, though. What do you think about this version? Simplifies the thread a lot!
![image](https://github.com/user-attachments/assets/f3152dba-f894-47c6-9cd1-26ca7ac768ea)

Lyrcaxis commented:
> What do you think about this version? Simplifies the thread a lot! ![image](https://private-user-images.githubusercontent.com/32474602/412547132-f3152dba-f894-47c6-9cd1-26ca7ac768ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzODU4MTQsIm5iZiI6MTczOTM4NTUxNCwicGF0aCI6Ii8zMjQ3NDYwMi80MTI1NDcxMzItZjMxNTJkYmEtZjg5NC00N2M2LTljZDEtMjZjYTdhYzc2OGVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEyVDE4MzgzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhODE1Mzk5YjI4NmRlYzgyNDlhMWQ0NGU0OWM3ZGUxNDk1YmQ3MTc0OWNmM2JiN2ZkZDZkMjlmZDU4OTJiZmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gr8d7SBSmNCsnBzH14imBqBvnUXhF5G5BoYd9gTPS5Q)

Hmm I missed the fact that the thread would keep running if `queuedJobs` is not disposed.
.. and if **it is** disposed, there'd need to be a `try-catch` block, which I would rather avoid in this context.

martindevans commented:
Thanks for the review, I've pushed up some changes based on it.
 - Corrected formatting/naming
 - Catching cancellation in worker (there don't seem to be non-throwing versions of this for `BlockingCollection`)
 - Added back in the inner loop for multi-stage jobs

martindevans commented:
Tested with your suggested change to the Program.cs, cancelling now immediately terminates playback with no other side effects. The next attempt to call `EnqueueJob` throws an `ObjectDisposedException`.

Lyrcaxis commented on Core/KokoroEngine.cs:
```diff
@@ -8,42 +8,41 @@
+        cancellation.Cancel();
         foreach (var job in queuedJobs) { job.Cancel(); }
-        queuedJobs.Clear();
+        queuedJobs.Dispose();
```
This will still spawn an unhandled `ObjectDisposedException` when cancelling mid-job.

Could be resolved with something like this:
```cs
catch (Exception e) when (e is OperationCanceledException or ObjectDisposedException) { }
```
.. but this could theoretically swallow `ObjectDisposedExceptions` of other systems unintentionally.

Also, could move the `queueJobs.Dispose()` to `finally`, without side effects, but I think the simplest and most proper solution would be to just add the `hasExited` check on the outer loop as well.

martindevans replied:
An alternative idea, how about if `Dispose` simply calls `CompleteAdding` (which signals that the queue is finished, nothing more will be added) and then the worker thread cancels all of the remaining jobs and disposes the queue? I think that's maybe cleaner, calling `Dispose` just signals to the worker thread to do the disposal work and we don't have to worry about synchronisation too much.

The only issue I can see is that moves cancellation across to the worker thread, is that valid?

Lyrcaxis replied:
I just tried it and it works flawlessly! Looks like the cleanest option indeed; no cancellation tokens, no try-catch..

And cancellation of jobs is handled with `foreach (var job in queuedJobs) { job.Cancel(); }` no problem.

Lyrcaxis replied:
Also the `ThrowIf` can be discarded since `queuedJobs.Add()` will throw properly.

martindevans replied:
I've pushed up something based on this. Cancellation token is still there to interrupt taking and to cancel the inner loop.

Lyrcaxis reviewed (changes requested):
And thank you for the PR! There's one final small issue with `.Dispose()`.

Lyrcaxis commented on Core/KokoroEngine.cs:
```diff
@@ -8,42 +8,43 @@
-                        await Task.Delay(1);
-                    }
-                }
+        new Thread(() =>
+        {
+            try {
+                while (queuedJobs.TryTake(out var kokoroJob, -1, cancellation.Token))
+                    while (!cancellation.IsCancellationRequested && !kokoroJob.isDone)
+                        kokoroJob.Progress(model);
+            }
+            catch (OperationCanceledException) {
+                // If cancellation was requested just swallow the cancellation exception and exit.
             }
+
+            foreach (var kokoroJob in queuedJobs.GetConsumingEnumerable())
+                kokoroJob.Cancel();
+            queuedJobs.Dispose();
         }).Start();
```
No longer need the token! (+ minor style change)

```suggestion
        new Thread(() => {
            foreach (var kokoroJob in queuedJobs.GetConsumingEnumerable()) {
                while (!kokoroJob.isDone) { kokoroJob.Progress(model); }
            }
        }).Start();
```
// optional `queuedJobs.Dispose()`, but it should be empty by thread exit and properly collected anyway!

martindevans replied:
I think you do need the token. As I understand it `CompleteAdding` just marks the collection as complete, so the enumerable will end once it has consumed **all** items. That's a change in behaviour from currently, where it will exit as soon as cancelled.

martindevans replied:
Oh I guess you're depending on the `Dipose` call calling cancel? So the worker will pull those jobs off the queue, but will immediately exit because they've been cancelled. That makes sense.

Lyrcaxis replied:
Yes, the original code had 2x `hasExited` checks which made it not obvious, but the inner one wasn't really needed.
Meanwhile I just love improvements like this:
![image](https://github.com/user-attachments/assets/c5172af3-b51a-47dc-bb9d-93e08081085f)

Lyrcaxis commented on Core/KokoroEngine.cs:
```diff
@@ -8,42 +8,43 @@
-        hasExited = true;
-        foreach (var job in queuedJobs) { job.Cancel(); }
-        queuedJobs.Clear();
+        queuedJobs.CompleteAdding();
+        cancellation.Cancel();
         model.Dispose();
         GC.SuppressFinalize(this);
```
Now with `CompleteAdding`, can do:
```suggestion
        foreach (var job in queuedJobs) { job.Cancel(); }
        queuedJobs.CompleteAdding();
        model.Dispose();
```
and handle the above without the cancellation token!

Lyrcaxis commented on Core/KokoroEngine.cs:
```diff
@@ -8,42 +8,43 @@
-    protected readonly ConcurrentQueue<KokoroJob> queuedJobs = [];
-
-    volatile bool hasExited;
+    protected readonly BlockingCollection<KokoroJob> queuedJobs = [];
+    readonly CancellationTokenSource cancellation = new();
```
```suggestion
    protected readonly BlockingCollection<KokoroJob> queuedJobs = [];
```

Lyrcaxis reviewed (approved):
Awesome, thanks for your contribution!

`CompleteAdding` turns out to be a blessing for this! I mention some slight changes/preferences below.
